### PR TITLE
Spark runner upgrade to Spark 2.0

### PR DESCRIPTION
**Pull request by amitsela**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:
- [ ] Make sure the PR title is formatted like:
  `[BEAM-<Jira issue #>] Description of pull request`
- [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
     Travis-CI on your fork and ensure the whole test matrix passes).
- [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
     number, if there is one.
- [x] If this contribution is large, please file an Apache
     [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

Spark 2.0.0-SNAPSHOT dependency, spark-core is provided by spark-sql now

Spark 2.0 uses Iterator instead of Iterable

Accumulators are compared on "zero" so using null safely is easier than implementing equals() and
hashCode() for NamedAggregators

Catch and log InterruptedException in close()

### CODE REVIEW

**Comment by amitsela** [comment]
R:@jbonofre 
First step towards integrating with Spark 2.0, mainly dependencies and backward compatibility (runner still uses RDD/DStream API).

**Comment by jbonofre** [comment]
Starting the review.

**Comment by amitsela** [comment]
### Notes for latest commits:

- Decoupled EvaluationContext from implementation - this allows to keep using existing streaming implementation (until migrated to Dataset API as well) while changing the batch implementation to Datasets.

- Encoders currently using kryo and double casting to easily support generics. Once Encoder API is released (2.1 ?) it'll probably be better to implement Encoders for common Beam Objects such as `WindowedValue` and `KV`.

- Migrated from SparkContext to SparkSession (2.0)

- `Accumulators` (`Aggergator` implementation in Spark) are currently backward-compatible - it'd be nice to upgrade to `AccumulatorV2`.

- Translations for batch transformations to the Dataset API are complete, excluding IOs - this is a WIP, but I'd like to share my thoughts here, I think that implementing `Read.from` as intended in https://issues.apache.org/jira/browse/BEAM-17 is possible only with the RDD API, and that's OK (RDD-Dataset are interoperable) and higher-level transformations such as `TextIO` will be implemented as direct translation to `DataFrameReader.text` which provide further optimizations by the Spark engine.
**_Note: all failing tests are because of the lack of IO transformations support._**  

- Concerning `GroupByKey` transformation, since it expects an `Iterable` output value, this implementation materializes the data which might be risky (OOM) but that's the current Beam model as well (which might be better of using `Iterator` instead). All `Combine` implementations (except for `globally`) will use the `combinePerKey` implementation so this will be a problem only if the user insists on grouping by key and applying a transformation to the grouped elements. Maybe this could be optimized as well by holding on to the `KeyValueGroupedDataset` in the context..

Finally, this is a WIP and I plan to work towards resolving IOs first, starting from `Read.from` and moving to higher level IOs. Once the runner supports batch in full, I think we could consider merging with master.

**Comment by amitsela** [comment]
@jbonofre @tomwhite @kennknowles @dhalperi please review, I'd be happy to hear your thoughts on my notes to this branch as well.
Thanks!

BTW, we should consider adding [WIP] tag to PRs so that Jenkins/Travis won't try to build WIP PRs. Apache Spark work this way and it makes sense.

**Comment by dhalperi** [comment]
>BTW, we should consider adding [WIP] tag to PRs so that Jenkins/Travis won't try to build WIP PRs. Apache Spark works this way and it makes sense.

Uh, that would be awesome.

**Comment by amitsela** [comment]
Sorry about the rebase.. the branch was outdated. Of course I will cherry-pick into a new PR later.

**Comment by kennknowles** [comment]
Perhaps it would make sense to merge `apache:master` into `apache:runners-spark2`? (via PR of course :-)

**Comment by amitsela** [comment]
Thanks @kennknowles !
After https://github.com/apache/incubator-beam/pull/585 is done I'll cherry-pick my changes, close this PR and open a new one. Sorry for the trouble.

**Comment by dhalperi** [comment]
#585 is now merged into branch, maybe time for a rebase?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkKeyedCombineFn.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
+package org.apache.beam.runners.spark.translation;
```
Comment (dhalperi) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkKeyedCombineFn.java]: Looks like this file is missing Apache license header, so RAT is complaining in the build.

Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkKeyedCombineFn.java]: Ah yes.. I'll update this in the next wave of commits, Thanks!

**Comment by jbonofre** [comment]
Resuming review.

**Comment by amitsela** [comment]
I just pushed basic support for Read.Bound which basically seems to support IOs without specific implementations (except for HadoopIO which is a runner-specific transformation).
For Bounded pipelines, I would say that this is still missing higher-level, optimized, translations for IOs that should be added by 1-to-1 mapping of BoundedSource to DataFrameReader (where possible) or InputFormat (for example SequenceFiles).

If anyone sees a better low-level entry-point to creating RDDs/DataFrames(sets) from Sources please let me know.

**Comment by amitsela** [comment]
Added support for Combine.Globally + an ongoing update against Spark's Aggregator.
R: @kennknowles I've exposed the fn in `Combine.Globally`

**Comment by kennknowles** [comment]
(sorry for the delay in review)

Exposing the `CombineFn` in `Combine.Globally` is fine by me. The fundamental semantic data that makes up a transform is very reasonable to expose, and we've done the same whenever we needed it to create an adequate override. LGTM to that.

**Comment by wonook** [comment]
Hey guys! I've been following through this PR, and I'm just wondering if there's an estimation of when this PR might be merged. It seems like it's been put on hold for quite a while 😕

**Comment by jbonofre** [comment]
We are still working on it (update and also better IO support).

**Comment by amitsela** [comment]
@wonook I've been working on this for a while, but unfortunately, the release of Spark 2.0 - Structured Streaming - was presented as Alpha and NOT production ready.
I'm still waiting on the following to be able to move forward:
- Purging of "old" time windows - for now, everything stays forever.. which makes it impossible to run long running (event-time) aggregations.
- State Store API.
- Encoder API.
I will keep this PR updated, so keep following 😉

**Comment by kennknowles** [comment]
R: -@kennknowles

Taking myself off the immediate list of reviewers, mostly because it isn't actively under review. Feel free to re-add me any time.

**Comment by davorbonaci** [comment]
Should we close this if it is not active, and track it in JIRA instead, which can link back to the closed PR? Or, alternatively, can it be merged to a feature branch?

**Comment by amitsela** [comment]
Closing, I will track in JIRA explaining ALL considerations to be taken if we would want to enable support for Spark 2.x

**Comment by yuchaoran2011** [comment]
@amitsela Did you create a jira issue for upgrading to Spark 2.0/2.1? I didn't find the relevant ticket

**Comment by jbonofre** [comment]
@yuchaoran2011 yes I created the Jira and I have a PR almost ready.

**Pull request closed without merging**